src/Functions/EntropyLearnedHash.cpp

```diff
@@ -0,0 +1,291 @@
+#include <base/defines.h>
+#include <base/types.h>
+#include <Columns/ColumnsNumber.h>
+#include <Columns/ColumnString.h>
+#include <Common/Exception.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/IFunction.h>
+#include <Interpreters/Context.h>
+
+/// Implementation of entropy-learned hashing: https://doi.org/10.1145/3514221.3517894
+/// If you change something in this file, please don't deviate too much from the pseudocode in the paper!
+
+/// TODOs for future work:
+/// - allow to specify an arbitrary hash function (currently always CityHash is used)
+/// - allow function chaining a la entropyLearnedHash(trainEntropyLearnedHash())
+/// - support more datatypes for data (besides String)
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int ILLEGAL_COLUMN;
+}
+
+namespace
+{
+
+using Key = String;
+using PartialKeyPositions = std::vector<size_t>;
+using Entropies = std::vector<size_t>;
+
+Key getPartialKey(std::string_view key, const PartialKeyPositions & partial_key_positions, Key & result)
+{
+    result.clear();
+    result.reserve(partial_key_positions.size());
+
+    for (auto partial_key_position : partial_key_positions)
+        if (partial_key_position < key.size())
+            result.push_back(key[partial_key_position]);
+
+    return result;
+}
+
+bool allPartialKeysAreUnique(const std::vector<Key> & data, const PartialKeyPositions & partial_key_positions)
```

Comment:
it should be equivalent to smth like "on the last nextBytes() we had 0 collisions"?

Reply:
Yepp, that's the intuitive meaning.